### PR TITLE
Allow modification of AccountInfo lamports field

### PR DESCRIPTION
**Pull request by LucasSte**
The field `lamports` of the struct `AccountInfo` is mutable, so we should allow its modification. For this we needed two modifications:

1. lamports is of type `Type::Ref(Type::Ref(Type::Uint(64)))` after accessing the struct member, we needed to include an extra load in sema do deal with the double pointer.
2. When we have `AccountInfo ai = tx.accounts[0];`, `ai` becomes a pointer to AccountInfo. The unused variable elimination was incorrectly eliminating references to structs, so I fixed such an issue.

### CODE REVIEW

File: src/codegen/unused_variable.rs [hunk 1]
```diff
@@ -87,6 +87,15 @@ pub fn should_remove_variable(pos: usize, func: &Function, opt: &Options) -> boo
         return true;
     }
 
+    // The unused variable elimination pass does not deal with pointers correctly, so we
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/codegen/unused_variable.rs]: The problem was elsewhere. Now, I've fixed it.

File: src/sema/expression/assign.rs [hunk 2]
```diff
@@ -116,6 +116,27 @@ pub(super) fn assign_single(
...
+                    left: Box::new(left),
+                    right: Box::new(val.cast(
+                        &right.loc(),
+                        inner.deref_memory(),
+                        true,
+                        ns,
+                        diagnostics,
+                    )?),
+                })
+            }
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/sema/expression/assign.rs]: I used `.cast()`. Check it looks better now.

**Comment by seanyoung** [comment]
> 1. lamports is of type `Type::Ref(Type::Ref(Type::Uint(64)))` after accessing the struct member, we needed to include an extra load in sema do deal with the double pointer.

I don't understand. This worked in solang v0.3.0, but regressed in v0.3.1. Why does it now have such a strange type?

**Comment by LucasSte** [comment]
> > 1. lamports is of type `Type::Ref(Type::Ref(Type::Uint(64)))` after accessing the struct member, we needed to include an extra load in sema do deal with the double pointer.
> 
> I don't understand. This worked in solang v0.3.0, but regressed in v0.3.1. Why does it now have such a strange type?

In PR #1325, we started loading readonly members in codegen, instead of emit. For that to work, I modified the layout of the builtin struct `AccountInfo` in `builtin_structs.rs` to match that of `solana_sdk.h`. This file specifies `uint64_t *`, as the type of lamports. After that we run the `struct_gep` LLVM instruction to retrieve such a field, the type we have is a `uint64_t **` (i.e. `Type::Ref(Type::Ref(Type::Uint(64)))`, which needs to be leaded before we manipulate it.

My reasoning for the changes here was to maintain the correspondence of layouts between sema and emit. Another solution would be to load the lamports field in emit, provided that we change the layout in sema.

> This worked in solang v0.3.0, but regressed in v0.3.1.

Although the compiler did not raise any error in v0.3.0, it wasn't properly working because the unused variable elimination was incorrectly removing assignments to struct references. This PR rectifies this mistake.

**Review by seanyoung — APPROVED** [review]

File: src/sema/expression/assign.rs [hunk 3]
```diff
@@ -360,6 +373,22 @@ pub(super) fn assign_expr(
                         diagnostics,
                     )?),
                 }),
+                // If the variable is a Type::Ref(Type::Ref(..)), we must load it first.
+                Type::Ref(inner)
```
Comment (xermicus) [inline src/sema/expression/assign.rs:377]: Same

**Review by xermicus — APPROVED** [review]

File: src/sema/expression/assign.rs [hunk 4]
```diff
@@ -116,6 +116,19 @@ pub(super) fn assign_single(
             right: Box::new(val.cast(&right.loc(), var_ty, true, ns, diagnostics)?),
         }),
         _ => match &var_ty {
+            // If the variable is a Type::Ref(Type::Ref(..)), we must load it.
+            Type::Ref(inner) if matches!(**inner, Type::Ref(_)) => Ok(Expression::Assign {
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/sema/expression/assign.rs:120]: How would you use it here?

[the remark this answers is not in the document]
Comment (xermicus) [reply src/sema/expression/assign.rs:120]: Didn't try it out, but by my understanding `match &var_ty.deref_memory() { Type::Ref(_) } ` should be the same as  `match &var_ty { Type::Ref(inner) if matches!(**inner, Type::Ref(_)) ` but more readable

[the remark this answers is not in the document]
Comment (LucasSte) [reply src/sema/expression/assign.rs:120]: I tried it, but it is not possible. In this match statement, we need to know whether we have a reference or not. If the type is not a reference, the expression is not assignable.

[the remark this answers is not in the document]
Comment (xermicus) [reply src/sema/expression/assign.rs:120]: Ah I see you need both types in the match arm

**Pull request merged**